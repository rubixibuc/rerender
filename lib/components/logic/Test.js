import React from "react";
import PropTypes from "prop-types";

const render = ({ children, value }) => (
  <React.Fragment>{children}</React.Fragment>
);

render.propTypes = {
  children: PropTypes.node,
  value: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.bool,
    PropTypes.number,
    PropTypes.object,
    PropTypes.string,
    PropTypes.func
  ])
};

render.displayName = "Rerender.Test";

export default render;
