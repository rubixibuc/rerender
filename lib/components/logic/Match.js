import React from "react";
import PropTypes from "prop-types";

import Test from "./Test";

const render = ({ children, first, matcher, expression }) => {
  let met = false;

  return children
    ? React.Children.map(children, child => {
        if (
          !met &&
          matcher(
            typeof child.props.value === "function"
              ? child.props.value()
              : child.props.value,
            typeof expression === "function" ? expression() : expression
          )
        ) {
          met = first && true;
          return child;
        }
        return null;
      })
    : null;
};

const TestPropType = (props, propName, componentName, location, propFullName) =>
  props[propName].type !== Test
    ? new Error(
        "Invalid prop `" +
          propFullName +
          "` supplied to" +
          " `" +
          componentName +
          "`. Validation failed."
      )
    : undefined;

render.propTypes = {
  children: PropTypes.oneOfType([
    TestPropType,
    PropTypes.arrayOf(TestPropType)
  ]),
  first: PropTypes.bool,
  matcher: PropTypes.func,
  expression: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.number,
    PropTypes.object,
    PropTypes.string,
    PropTypes.func
  ])
};

render.defaultProps = {
  first: true,
  matcher: (a, b) => a === b
};

render.displayName = "Rerender(Match)";

export default render;
