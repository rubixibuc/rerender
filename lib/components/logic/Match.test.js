import React from "react";
import renderer from "react-test-renderer";

import Match from "./Match";
import Test from "./Test";

test("should render nothing when children are undefined", () => {
  const component = renderer.create(<Match expression={0} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render Tests with matching value", () => {
  const component = renderer.create(
    <Match expression={2 + 2}>
      <Test value={4}>4</Test>
      <Test value={2}>2</Test>
      <Test value={"4"}>'4'</Test>
      <Test value={4}>4</Test>
    </Match>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render Tests with matching value when value is function", () => {
  const component = renderer.create(
    <Match expression={2 + 2}>
      <Test value={() => 4}>4</Test>
      <Test value={2}>2</Test>
    </Match>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
