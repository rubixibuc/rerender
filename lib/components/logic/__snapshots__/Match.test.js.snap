// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Tests with matching value when value is function 1`] = `"4"`;

exports[`should render all Tests (1, 4) with matching value 1`] = `
Array [
  "1",
  "4",
]
`;

exports[`should render matching Test (1) when Match expression is function 1`] = `"1"`;

exports[`should render nothing when children are undefined 1`] = `null`;

exports[`should render only first Test (1) with matching value 1`] = `"1"`;
